Add some symbols for retrieving locale info

utelle opened the pull request:
The following symbols were added:

- wxLOCALE_MEASURE_METRIC : check whether locale uses metric measurements
- wxLOCALE_CURRENCY_SYMBOL : the currency symbol (like "$")
- wxLOCALE_CURRENCY_CODE : the ISO currency code (like "USD" or "EUR")
- wxLOCALE_CURRENCY_DIGITS : number of currency digits

Documentation and test cases are included.

utelle committed:
Add some symbols for retrieving locale info

- wxLOCALE_MEASURE_METRIC : check whether locale uses metric measurements
- wxLOCALE_CURRENCY_SYMBOL : the currency symbol (like "$")
- wxLOCALE_CURRENCY_CODE : the ISO currency code (like "USD" or "EUR")
- wxLOCALE_CURRENCY_DIGITS : number of currency digits

utelle committed:
Adjust implementations and add documentation

- added test cases
- minor change for MSW
- completed Unix implementation
- added implementation for wxLocale for Unix

utelle committed:
Remove trailing whitespace

utelle committed:
Fix trailing whitespace (macOS)

utelle committed:
Fix wrong variable name in macOS implementation

oneeyeman1 commented:
@utelle ,
Thx

utelle committed:
Add missing closing parenthesis

vadz commented on src/msw/uilocale.cpp:
```diff
@@ -481,6 +481,26 @@ class wxUILocaleImplName : public wxUILocaleImpl
 
+            case wxLOCALE_MEASURE_METRIC:
+                str = DoGetInfo(LOCALE_IMEASURE);
+                if (str.IsEmpty())
```
We prefer using standard-like functions:
```suggestion
                if (str.empty())
```

vadz commented on src/unix/uilocale.cpp:
```diff
@@ -594,6 +594,48 @@ wxUILocaleImplUnix::GetInfo(wxLocaleInfo index, wxLocaleCategory cat) const
+        {
+            wxString measureStr = GetLangInfo(_NL_MEASUREMENT_MEASUREMENT);
+            if (measureStr.IsEmpty())
+                measureStr = wxString("\0");
```
This return value is not documented and it's not clear what does it correspond to.

vadz commented on include/wx/localedefs.h:
```diff
@@ -89,6 +89,19 @@ enum wxLocaleInfo
     wxLOCALE_DECIMAL_POINT,
 
+    // check whether locale uses the metric system for measurements
+    // returns "Yes", "No", or "Unknown"
```
I'd probably rather use "1", "0" and "" instead of these strings.

In any case, let's define `wxLOCALE_TRUE`, `_FALSE` and `_UNKNOWN` constants for them instead of using the literal strings.

Alternatively, perhaps it would be better to have a `wxLocaleMeasurement` enum and a separate function `GetMeasurementSystem()` returning it? This would be much cleaner IMO.

vadz commented on include/wx/localedefs.h:
```diff
@@ -89,6 +89,19 @@ enum wxLocaleInfo
+    // returns "Yes", "No", or "Unknown"
+    wxLOCALE_MEASURE_METRIC,
+
+    // the currency symbol (for example "$")
+    wxLOCALE_CURRENCY_SYMBOL,
+
+    // the currency ISO code (for example "USD")
+    wxLOCALE_CURRENCY_CODE,
+
+    // the number of currency digits
+    wxLOCALE_CURRENCY_DIGITS,
```
We could also define `wxLocaleCurrencyInfo` struct with `int digits` field which would avoid to have to return this information as a string.

vadz reviewed (changes requested):
Thanks, having access to this information definitely seems useful, but it doesn't seem to fit `GetInfo()` very well. We can make it fit there if you prefer not to add the new functions but in this case I'd like to at least define constants for the various values.

Thanks again!

utelle commented:
> Thanks, having access to this information definitely seems useful, but it doesn't seem to fit `GetInfo()` very well. We can make it fit there if you prefer not to add the new functions but in this case I'd like to at least define constants for the various values.

I don't have any preferences. Using `GetInfo()` was the simplest approach to test the functionality under Windows and Unix. Unfortunately, I don't have access to a macOS system right now, so that I had to add the implementation based on information from the Apple documentation.

If you prefer additional methods for `wxUILocale`, I'm open for proposals, how to name them. The same new methods would have to be added to `wxLocale`.

I will look into your other comments later today.

utelle commented:
@vadz: Before looking into the implementation details I'd like to make some remarks about the failing CI runs.

There are 6 failing checks (while over 30 succeed - a better ratio than I had expected):

- [Cirrus CI / FreeBSD wxGTK3](https://github.com/wxWidgets/wxWidgets/pull/25765/checks?check_run_id=49764252622)
  Under FreeBSD most `nl_langinfo` symbols are not available. As a workaround, we could use `localeconv` (which should work). The drawback is that `localeconv` only works for the global locale.
- [CMake builds / macOS 14 wxOSX Xcode](https://github.com/wxWidgets/wxWidgets/actions/runs/17520773749/job/49764257927?pr=25765)
  Just like the other failing macOS runs an older macOS version 14.7.6 is used. For some reason a `wxLocale` method is used as a fallback (and in that method not all symbols are handled).
- [continuous-integration/appveyor](https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/52697298)
  The VS14 run fails, probably due to misinterpreting the €-sign. Maybe it's not such a good idea to encode it directly in the test source code. I'll try `"\u20AC"` instead.

- [Mac builds / wxMac ARM ASAN](https://github.com/wxWidgets/wxWidgets/actions/runs/17520773760/job/49764252808?pr=25765)
- [Mac builds / wxMac Intel C++17](https://github.com/wxWidgets/wxWidgets/actions/runs/17520773760/job/49764252806?pr=25765)
- [Mac builds / wxMac Universal C++14](https://github.com/wxWidgets/wxWidgets/actions/runs/17520773760/job/49764252802?pr=25765)

For the above 3 macOS runs it looks like that the locales are not properly installed on the runner, even the `en_US` locale returns the general currency symbol `"\u00A4"` instead of `"$"`.

Before I change anything in the API, I'd like to get rid of those failing CI runs.

Now, regarding the API I'd like to stick with `GetInfo`. Actually, the same system functions are used under the hood. And implementing of handling the new `wxLOCALE` symbols requires no changes to the API of `wxUILocale` and `wxLocale`. However, we could add a convenience method for retrieving all the currency information in one call and one data structure. Not sure, whether that convenience method should be added to `wxLocale` at all.

utelle committed:
Fix various issues

- use "\u20AC" for representing the Euro symbol in test cases
- adjust Unix implementation, if the operating system doesn't provide glibc
- implement missing cases in GetInfo for macOS

utelle committed:
Replace NULL by nullptr

utelle committed:
Fix macOS issues

- correct wrong variable name
- include missing header file

utelle committed:
Add missing type casts

utelle committed:
Correct retrieval of number of currency digits

vadz commented:
> @vadz: Before looking into the implementation details I'd like to make some remarks about the failing CI runs.
> 
> There are 6 failing checks (while over 30 succeed - a better ratio than I had expected):
> 
>     * [Cirrus CI / FreeBSD wxGTK3](https://github.com/wxWidgets/wxWidgets/pull/25765/checks?check_run_id=49764252622)
>       Under FreeBSD most `nl_langinfo` symbols are not available. As a workaround, we could use `localeconv` (which should work). The drawback is that `localeconv` only works for the global locale.

Yes, the symbols you use are not POSIX (see https://pubs.opengroup.org/onlinepubs/009695399/basedefs/langinfo.h.html for the list of the standard ones), so you have to either add specific checks for them to configure/CMake or use them only for glibc, which is known to have them (assuming nobody else does). Note that POSIX does have `CRNCYSTR`, so we probably should use it instead of `CURRENCY_SYMBOL` (which is a glibc extension).

>     * [CMake builds / macOS 14 wxOSX Xcode](https://github.com/wxWidgets/wxWidgets/actions/runs/17520773749/job/49764257927?pr=25765)
>       Just like the other failing macOS runs an older macOS version 14.7.6 is used. For some reason a `wxLocale` method is used as a fallback (and in that method not all symbols are handled).

It's ok to make the test conditional on the functionality being actually supported, I think.

>     * [continuous-integration/appveyor](https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/52697298)
>       The VS14 run fails, probably due to misinterpreting the €-sign. Maybe it's not such a good idea to encode it directly in the test source code. I'll try `"\u20AC"` instead.

I think we use `/utf-8` now, don't we? If so, this should work... But I don't have this version of MSVS to test with any longer.

>     * [Mac builds / wxMac ARM ASAN](https://github.com/wxWidgets/wxWidgets/actions/runs/17520773760/job/49764252808?pr=25765)
> 
>     * [Mac builds / wxMac Intel C++17](https://github.com/wxWidgets/wxWidgets/actions/runs/17520773760/job/49764252806?pr=25765)
> 
>     * [Mac builds / wxMac Universal C++14](https://github.com/wxWidgets/wxWidgets/actions/runs/17520773760/job/49764252802?pr=25765)
> 
> 
> For the above 3 macOS runs it looks like that the locales are not properly installed on the runner, even the `en_US` locale returns the general currency symbol `"\u00A4"` instead of `"$"`.

This character is probably used for "C" locale, so maybe `en_US` is not even supported there?

> Before I change anything in the API, I'd like to get rid of those failing CI runs.

Sure.

> Now, regarding the API I'd like to stick with `GetInfo`. Actually, the same system functions are used under the hood.

I don't think this is a good argument. The user of the API doesn't really care about how it is implemented. And returning integers as strings is rather gross.

> And implementing of handling the new `wxLOCALE` symbols requires no changes to the API of `wxUILocale` and `wxLocale`. However, we could add a convenience method for retrieving all the currency information in one call and one data structure.

Yes, this is what I proposed with `wxLocaleCurrencyInfo wxUILocale::GetCurrencyInfo()` in my first comment.

> Not sure, whether that convenience method should be added to `wxLocale` at all.

I don't think it needs to be, it's fine to add new functionality only to `wxUILocale` IMO.

utelle commented:
> > * [Cirrus CI / FreeBSD wxGTK3](https://github.com/wxWidgets/wxWidgets/pull/25765/checks?check_run_id=49783355942)
> >   Under FreeBSD most `nl_langinfo` symbols are not available. As a workaround, we could use `localeconv` (which should work). The drawback is that `localeconv` only works for the global locale.
> 
> Yes, the symbols you use are not POSIX (see https://pubs.opengroup.org/onlinepubs/009695399/basedefs/langinfo.h.html for the list of the standard ones), so you have to either add specific checks for them to configure/CMake or use them only for glibc, which is known to have them (assuming nobody else does). Note that POSIX does have `CRNCYSTR`, so we probably should use it instead of `CURRENCY_SYMBOL` (which is a glibc extension).

Except for `CRNCYSTR` there is no POSIX equivalent for the other symbols added by this PR. However, using `localeconv()` allows to retrieve the values in question. And if we use `TempLocaleSetter` as in the `wxUILocaleImplUnix::GetLangInfo()` method in conjunction with `localeconv`, it should work for the general case in `wxUILocale`, too.

At the moment, the global static method `wxLocale::GetInfo()` (without `TempLocaleSetter`) is used as a fallback, if the symbols are not available. However, I intend to change that.

The reason that this CI run still fails is that I forgot to check for symbol availability in one place. All other Unix CI runs now already succeed.

> > [CMake builds / macOS 14 wxOSX Xcode](https://github.com/wxWidgets/wxWidgets/actions/runs/17528995525/job/49783349882?pr=25765)
> 
> It's ok to make the test conditional on the functionality being actually supported, I think.

This CI run still fails for the test cases using English and German locales via `wxUILocale`. Interestingly, the `wxLocale::GetInfo` tests I added based on a French locale do **not** fail. However, I don't know whether this is so, because the French locale is - for whatever reason - properly installed, or because different functions are used to retrieve the values.

> > * continuous-integration/appveyor
> >   The VS14 run fails, probably due to misinterpreting the €-sign. Maybe it's not such a good idea to encode it directly in the test source code. I'll try `"\u20AC"` instead.
> 
> I think we use `/utf-8` now, don't we? If so, this should work... But I don't have this version of MSVS to test with any longer.

I made the change as described above. And it works. This CI run no longer fails.

> > * [Mac builds / wxMac ARM ASAN](https://github.com/wxWidgets/wxWidgets/actions/runs/17528995533/job/49783341748?pr=25765)
> > * [Mac builds / wxMac Intel C++17](https://github.com/wxWidgets/wxWidgets/actions/runs/17528995533/job/49783341741?pr=25765)
> > * [Mac builds / wxMac Universal C++14](https://github.com/wxWidgets/wxWidgets/actions/runs/17528995533/job/49783341737?pr=25765)
> >   
> > For the above 3 macOS runs it looks like that the locales are not properly installed on the runner, even the `en_US` locale returns the general currency symbol `"\u00A4"` instead of `"$"`.
> 
> This character is probably used for "C" locale, so maybe `en_US` is not even supported there?

Just as for the CI run [CMake builds / macOS 14 wxOSX Xcode](https://github.com/wxWidgets/wxWidgets/actions/runs/17528995525/job/49783349882?pr=25765) only the `wxUILocale` tests fail, not the `wxLocale` test (using French locale). I will add `wxUILocale` tests with French locale to check whether this behaviour is due the actual locale used or due to different functions used.

> > Now, regarding the API I'd like to stick with `GetInfo`. Actually, the same system functions are used under the hood.
> 
> I don't think this is a good argument. The user of the API doesn't really care about how it is implemented. And returning integers as strings is rather gross.

Yes, I agree, that isn't very elegant. Therefore a convenience method returning a structure with all curreny infos is certainly a good idea.

> > And implementing of handling the new `wxLOCALE` symbols requires no changes to the API of `wxUILocale` and `wxLocale`. However, we could add a convenience method for retrieving all the currency information in one call and one data structure.
> 
> Yes, this is what I proposed with `wxLocaleCurrencyInfo wxUILocale::GetCurrencyInfo()` in my first comment.

I intend to follow that route.

However, I'm still not sure how to handle `wxLocale`. Currently I use `wxLocale::GetInfo()` as a fallback method. This is convenient and avoids code duplication. And on top adds the same functionality to `wxLocale`.

> > Not sure, whether that convenience method should be added to `wxLocale` at all.
> 
> I don't think it needs to be, it's fine to add new functionality only to `wxUILocale` IMO.

The question is whether you can get used to keeping the additional `wxLOCALE_` symbols. Or whether you prefer just 2 additional methods, one for the curreny values (symbol, code, number of fractional digits) and one for determining whether the measurement system is metric or not.

utelle committed:
Add test with French locale, add missing check

utelle committed:
Correct test of French locale

vadz commented:
> Or whether you prefer just 2 additional methods, one for the curreny values (symbol, code, number of fractional digits) and one for determining whether the measurement system is metric or not.

I think I'd prefer to have just the functions. They can share the implementation with `GetInfo()` internally, of course, if this is convenient (although I don't believe it really helps that much).

utelle committed:
Use locale with regions for currency tests

utelle commented:
> > Or whether you prefer just 2 additional methods, one for the curreny values (symbol, code, number of fractional digits) and one for determining whether the measurement system is metric or not.
> 
> I think I'd prefer to have just the functions. They can share the implementation with `GetInfo()` internally, of course, if this is convenient (although I don't believe it really helps that much).

Ok, I will then start to change the implementation in that direction.

BTW, I found out what was causing the failing macOS tests: the tests instantiated locales for `en`, `de`, and `fr`. While Windows and Linux seem to associate a default region to the locale in that case, macOS does not. If asked for locale `fr`, you get a locale **without** an associated region. Without region, macOS does not know which currency to select and returns the general currency symbol (&#x00A4; == `\00A4`) and no currency code.

After I changed the tests to instantiate locales `en-US`, `de-DE`, and `fr-FR`, the tests succeeded.

The FreeBSD CI run still fails, but now on running the test. The results returned for currency look strange. At the moment I have no idea, what's going.

utelle committed:
Rework wxUILocale API (number/currency formatting)

The additional wxLOCALE_ symbols for retrieving information via wxUILocale::GetInfo were removed and replaced by the following new wxUILocale API methods:

- GetNumberFormatting : retrieve information relevant for formatting ordinary numbers
- GetCurrencySymbol : retrieve the currency symbol
- GetCurrencyCode : retrieve the international 3-letter currency code
- GetCurrencySymbolPosition : retrieve the position of the currency symbol
- GetCurrencyInfo : retrieve currency information, including formatting parameters
- UsesMetricSystem : determine whether the measurement system is metric or not

utelle committed:
Handle unused method, include different macOS header

utelle committed:
Remove keyword override before method body (macOS)

utelle committed:
Fix macOS syntax error and class prefix

utelle committed:
Add override keyword

utelle committed:
Add missing header <array>, fix method names

utelle commented:
@vadz: I have reworked the wxUILocale API. In the end I removed all new `GetInfo` symbols and created separate methods to access the currency information. 

Additionally, there is now a method to retrieve the information regarding locale-depending formatting of ordinary numbers. This method retrieves all relevant data in one go - especially on macOS this is efficient, because it requires to instantiate a NSNumberFormatter only once.

2 GitHub CI runs currently fail:

- [CMake builds / MSW/MSVC wxQt 5.15](https://github.com/wxWidgets/wxWidgets/actions/runs/17627523048/job/50087913317?pr=25765) - a GUI test fails - IMHO not related to my changes
- [Cirrus CI / FreeBSD wxGTK 3](https://github.com/wxWidgets/wxWidgets/pull/25765/checks?check_run_id=50087913183) - all currency tests fail.  
  It's unclear why the tests fail. And it's difficult to find out, because I don't have access to a FreeBSD system. I started to set up a FreeBSD VM, but that is a major task I already spent several hours on.

Nevertheless, I'd like to ask you to review the PR again. If you agree with the API, I will add documentation for the new methods.

vadz commented:
P.S. And the other failure is #25779 i.e. indeed spurious.

vadz commented:
> Is there any documentation available about this feature, and how to actually debug an application. I would need to see, what data exactly are retrieved from the system.

If you have access to it (perhaps you don't?) just select "Re-Run with Terminal Access" from the button menu:

<img width="762" height="350" alt="image" src="https://github.com/user-attachments/assets/9d0af2d7-a360-497b-b959-3335e1f9663c" />

and you'll be dropped (once the build starts) into a ssh session on the build machine, from where you can do anything you want, e.g. debug, compile and run test code etc.

utelle committed:
Use LC_ALL instead of LC_CTYPE

For Unix system without glibc the function localeconv is used to retrieve locale-dependent data. However, the underlying data structure is filled on partially when LC_CTYPE is used for setting the temporary locale.

utelle commented:
@vadz:

Locally, I adjusted the class `wxNumberFormatter` to take the grouping of digits into account for applying the _grouping separator_ aka _thousands separator_. No change to the public API so far, because IMHO distinguishing between _thousands separator_ and _grouping separator_ makes no sense - in effect they are synonyms.

The question is whether to make adjustments to `wxNumberFormatter` part of this PR or not.

Regarding `wxNumberFormatter` I would like to add style flags to explicitly format _currency values_:

- `Style_Currency` : currency value without currency symbol
- `Style_Currency_Local` : currency value with local currency symbol (prefix or suffix according to local conventions)
- `Style_Currency_International` : currency value with international currency symbol aka ISO4217 currency code (always prefix)

Tests could be adopted more or less from PR https://github.com/wxWidgets/wxWidgets/pull/1781.

vadz commented on tests/intl/intltest.cpp:
```diff
@@ -417,11 +417,35 @@ TEST_CASE("wxUILocale::IsSupported", "[uilocale]")
+    if (CheckSupported(locDE, "German"))
+    {
+        CHECK( locDE.GetInfo(wxLOCALE_DECIMAL_POINT) == ",");
+        CHECK( locDE.GetCurrencySymbol() == L"\u20AC");
```
I'm almost sure that using
```suggestion
        CHECK( locDE.GetCurrencySymbol() == L"€");
```
should work everywhere nowadays.

vadz commented on interface/wx/intl.h:
I think this is out of date now?

vadz reviewed (commented):
Thanks for all your work on this and sorry for the belated review!

Globally, the least clear thing to me (partially because it is not covered by the tests) is how do we handle digits grouping. AFAICS you use Win32 format for this string which is perhaps less confusing than POSIX one but still not great. I wonder why we can't represent it with a `vector<int>` containing the group sizes, with some special value (0? -1?) meaning "repeat last size". This is obviously less efficient than a simple string, but we don't care about efficiency that much and I think it would be much more clear.

Also, after looking at the code and the tests, I think that mayb you were right and we only need to have one public function, after all, returning all information at once instead of separate functions for separate parts of it. If you (still?) think this is better, please feel free to only export this function, I believe it would be simpler for you too — and there would be just one function to document, too.

Thanks again!

utelle commented on include/wx/localedefs.h:
```diff
@@ -122,6 +122,53 @@ enum wxLocaleForm
 
+enum class wxMeasurementSystem
+{
+    UnknownSystem,
```
> Minor, but I think `wxMeasurementSystem::UnknownSystem` is unnecessarily redundant, so I suggest just

Yes, I'll change that back to just `Unknown`. The name conflicted with another symbol, before I added the `class` attribute to the `enum`.

utelle commented on include/wx/localedefs.h:
```diff
@@ -122,6 +122,53 @@ enum wxLocaleForm
+    NonMetric
+};
+
+enum class wxCurrencySymbolPosition
```
> It's not really obvious whether having 4 values in the same enum or having 2 values and a separate `useSeparator` flag is better. I'd be curious to know why did you decide to make this choice?

Well, actually that is how the information is returned from the operating system under Windows. Under Linux/Unix there are 2 different property values. And under macOS the information has to be extracted manually from a formatted string.

I decided to define 4 enum values to be able to return the information in a single value that can be returned as a function value easily.

Currently, I have a separate function returning this value. If we decide to return the information within the currency structure only, then we could split the information into 2 separate values.

utelle commented on src/osx/core/uilocale.mm:
```diff
@@ -349,6 +362,131 @@ int CompareStrings(const wxString& lhs, const wxString& rhs,
+    NSString* formatted = [formatter stringFromNumber:@123];
+    NSString* symbol = formatter.currencySymbol;
+
+    NSRange symbolRange = [formatted rangeOfString:symbol];
```
> I think all this string manipulation stuff would be much more clear to an average wx programmer if it were written using `wxString` functions and not Cocoa ones... It's not worth rewriting it, of course, now that it has been already written like this.

I found it simpler to use Cocoa functions instead of converting the strings to wxStrings, because they are only used locally.

utelle commented on src/msw/uilocale.cpp:
```diff
@@ -642,6 +680,58 @@ class wxUILocaleImplName : public wxUILocaleImpl
+        wxString str = DoGetInfo(LOCALE_IMEASURE);
+        if (str.empty())
+        {
+            wxString region = GetLocaleId().GetRegion();
+            // In 2025 only in the United States, Liberia, and Myanmar
+            // the metric system is not the default measurement system.
+            bool usesMetric = !(region == "US" || region == "LR" || region == "MM");
+            return (usesMetric) ? wxMeasurementSystem::Metric : wxMeasurementSystem::NonMetric;
```
> I still would just return `Unknown` here. If you'd like, we could provide this code in a separate function called `GuessMetricSystemFromRegion()` or something like this. But I'd like to avoid providing wrong information to the program without it even knowing it that it could be wrong.

Only if the Windows API function `GetLocaleInfoEx()` fails, it would be necessary to deduce the metric system from the region. That should usually not happen at all. And if it fails, `wxLogLastError` will be called. So, the user will know that something went wrong.

Almost the whole world agreed on using the metric system, except for the 3 regions mentioned in the code. Therefore it is fair assumption to return `Metric`. macOS for example uses a boolean for this item, that is, you will never get `Unknown` on macOS.

Typically, an application would/should offer the user the choice of the measurement system, using the result of this function just as the default.

vadz replied:
`wxLogLastError()` is just a debugging helper, the program would never use it to know if anything went wrong.

I agree that returning `Unknown` would be rare, but it still seems more honest to do it than just trying to guess. And by using a separate function we would at least avoid having the same code in all 3 ports with all the usual maintenance issues due to it (what if Gibraltar becomes independent in 2027 and decides to use imperial system?).

utelle commented on src/osx/core/uilocale.mm:
```diff
@@ -349,6 +362,131 @@ int CompareStrings(const wxString& lhs, const wxString& rhs,
+{
+    wxCurrencySymbolPosition symbolPosition = wxCurrencySymbolPosition::PrefixWithSep;
+
+    NSNumberFormatter* formatter = [[NSNumberFormatter alloc] init];
```
> This has to be released somewhere. Typically this would be done by using `wxCFRef` for it.

Would it be ok to specify `autorelease`?

```
NSNumberFormatter* formatter = [[[NSNumberFormatter alloc] init] autorelease];
```

utelle replied:
I'm not experienced with macOS programming. However, if I include this header:

```
#import <objc/NSObject.h>
```

and then use

```
objc::strong_ptr<NSNumberFormatter> formatter([[NSNumberFormatter alloc] init]);
 ```

it should work, I guess.

utelle commented on src/osx/core/uilocale.mm:
```diff
@@ -349,6 +362,131 @@ int CompareStrings(const wxString& lhs, const wxString& rhs,
+        return (isMetric) ? wxMeasurementSystem::Metric : wxMeasurementSystem::NonMetric;
+    }
+    else
+    {
+        wxString region = GetLocaleId().GetRegion();
+        // In 2025 only in the United States, Liberia, and Myanmar
+        // the metric system is not the default measurement system.
+        bool usesMetric = !(region == "US" || region == "LR" || region == "MM");
+        return (usesMetric) ? wxMeasurementSystem::Metric : wxMeasurementSystem::NonMetric;
+    }
```
> No, this is not quite right, this method is only available since 10.12 according to [the docs](https://developer.apple.com/documentation/foundation/nslocale/usesmetricsystem?language=objc). This is still pretty old but we still support 10.10 officially, so we still need it.

Yes, you are right.

> And as I also argued before, it would really be best not to have this identical code snippet in 3 places.

I already added your proposed method for guessing the measurement system from the region. `Unknown` will be returned from `UsesMetricSystem()` if the information couldn't be retrieved reliably from the operating system.

utelle commented:
> > However, we will have to discuss, if and how `wxNumberFormatter` should support using a specific locale instead of just the current locale. And this may lead to API changes.
> 
> Maybe it's not worth doing this. `wxNumberFormatter` was supposed to be a simple way to format numbers in a GUI program and I think we may leave it like this. At the very least, I'd prefer to leave any big changes to it out of this PR.

IMHO the approach with these static `wxNumberFormatter` methods is highly inefficient, if a large number of values has to be formatted, because the formatting information has to be retrieved over and over again.

However, I agree that changes in that respect should be made in a separate PR.

> > Well, "international" is the term used elsewhere. "local" = within the country, "international" = across country borders.
> 
> It's an interesting point of view, but I don't think it's very common. Surely country codes are used in plenty of financial contexts "locally" too. I really think "currency symbol" and "currency code" are much simpler and completely unambiguous.

Well, names are just names. If it makes you happy,  we call them "currency symbol" and "currency code".

utelle committed:
Apply changes requested by review

utelle committed:
Fix macOS implementation, adjust test case

utelle committed:
Change initializer list to explicit assignment

Obviously not all compiler versions accept initializer lists for a struct without a matching constructor.

utelle committed:
Remove whitespace

utelle committed:
Remove whitespace

utelle committed:
Fix for compile errors on MSW and FreeBSD

utelle commented:
@vadz: 
Building wxWidgets now works in all CI jobs.

However, currently all **Mac builds** _that do perform the tests_ **fail** for one specific test case. Namely the test for checking different number formatting for ordinary numeric values and currency values. According to the Unicode CLDR database for the locale `de-AT` there are different _thousands separators_. But unfortunately not all systems follow the CLDR specifications. Especially, GLIBC defines the values for the _thousands_ separator almost the other way around than CLDR. Obviously, MacOS follows neither CLDR nor GLIBC.

The question is whether we should keep those tests at all. If we keep them, we need to check for MacOS. However, for me it is difficult to tell how this should be done.

Additionally, 2 AppVeyor runs fail during the tests for formatting according to Indian rules. Looking at the failing tests it seems that the Windows system on which the build takes place belong to those that return the wrong grouping information. A comment in the old PR https://github.com/wxWidgets/wxWidgets/pull/1781 reads:

>  ```C++
>    // On some Windows systems the grouping is "3;2" which is incorrect
>    // Skip this test in that case
> ```

I guess I have to check for that situation in the test case, too.

Nevertheless, I'd like to ask you to review the code of this PR once again.

P.S.: The documentation is not yet done. I will write it up as soon as the code is finalized.

vadz reviewed (commented):
Thanks for the updates!

Just a few comments left, please let me know if you disagree with them.

Concerning the failing tests, perhaps we could check that the locale functions return the expected data for the locale before continuing with the rest of the tests?

utelle commented on src/unix/uilocale.cpp:
```diff
@@ -1082,22 +1076,18 @@ wxUILocaleImplUnix::DoGetInfoFromLocaleConv(wxLocaleConvAttr index, wxLocaleCate
-            wchar_t thousandsSep[16];
             if (cat == wxLOCALE_CAT_MONEY)
-                mbstowcs(thousandsSep, lc->mon_thousands_sep, 16);
+                return wxString(lc->mon_thousands_sep, wxCSConv(GetCodeSet()));
```
> I still don't understand why do we need to use anything else than (default) `wxConvLibc` here, we've just changed locale to use `m_locId` so AFAICS it should work?

No, not in the general case. It would work for the locale set by `wxUILocale::UseDefault()`. However, you can instantiate other instances of `wxUILocale` without changing the C locale. And then the charset is not necessarily the same.

> If we really need to use a custom conversion, I'd construct it once before the `switch` instead of doing it in every `case`.

Well, the conversion will be done exactly once per call of `DoGetInfoFromLocaleConv`. That is, it will be constructed only once per call.

vadz replied:
> > I still don't understand why do we need to use anything else than (default) `wxConvLibc` here, we've just changed locale to use `m_locId` so AFAICS it should work?
> 
> No, not in the general case. It would work for the locale set by `wxUILocale::UseDefault()`. However, you can instantiate other instances of `wxUILocale` without changing the C locale. And then the charset is not necessarily the same.

Sorry to insist but there is `TempLocaleSetter` just above. Doesn't it set locale to exactly this locale?

> > If we really need to use a custom conversion, I'd construct it once before the `switch` instead of doing it in every `case`.
> 
> Well, the conversion will be done exactly once per call of `DoGetInfoFromLocaleConv`. That is, it will be constructed only once per call.

Yes, but it's still more code than just constructing it once.

vadz replied:
Note that `wxConvLibc` is implicit anyhow, so you don't need to use it explicitly (although you can do it, of course, if you find it more clear).

utelle commented on tests/intl/intltest.cpp:
```diff
@@ -417,33 +417,38 @@ TEST_CASE("wxUILocale::IsSupported", "[uilocale]")
-    CHECK( locEN.GetCurrencyInfo().CurrencyFormat.FractionalDigits == 2);
-    CHECK( locEN.GetCurrencySymbolPosition() == wxCurrencySymbolPosition::PrefixNoSep);
+    CHECK( locEN.GetCurrencyInfo().currencyFormat.fractionalDigits == 2);
+    CHECK( (currencyPosition == wxCurrencySymbolPosition::Prefix && !currencyHasSeparator) );
```
A single check was enough in my first implementation using 4 `wxCurrencySymbolPosition` values incorporating the use of a separator.

utelle reviewed (commented):
I will apply the changes soon, but most likely not before Monday.

utelle committed:
Apply changes from last review

utelle committed:
Fix typo in macOS code

vadz commented on include/wx/localedefs.h:
```diff
@@ -124,49 +126,32 @@ enum wxLocaleForm
-    int      FractionalDigits = 0;
+    wxString decimalSeparator;
+    wxString groupSeparator;
+    std::vector<size_t> grouping;
```
`size_t` was a mistake, basically all C++ experts recognize it by now (and "now" means at least a decade already). `wxString` has to use it for compatibility with `std::string`, but here we don't need to do it.

I think there will be fewer casts needed, if any, not more if you switch to `int` here.

vadz commented on src/unix/uilocale.cpp:
```diff
@@ -1056,13 +1045,18 @@ wxUILocaleImplUnix::DoGetNumberFormatting(wxLocaleCategory cat) const
 #endif
 
-    return wxLocaleNumberFormatting(groupSeparator, groupingStr, decimalSeparator, fractionalDigits);
+    wxLocaleNumberFormatting numForm;
+    numForm.decimalSeparator = decimalSeparator;
+    numForm.groupSeparator   = groupSeparator;
+    numForm.grouping         = grouping;
+    numForm.fractionalDigits = fractionalDigits;
+    return numForm;
```
Sorry for misleading you into removing the ctor :-( I guess if we can't use the initializer list, it should be restored.

vadz commented on include/wx/private/uilocale.h:
```diff
@@ -100,7 +100,7 @@ class wxUILocaleImpl : public wxRefCounter
     virtual wxString GetCurrencySymbol() const = 0;
     virtual wxString GetCurrencyCode() const = 0;
-    virtual wxCurrencySymbolPosition GetCurrencySymbolPosition() const = 0;
+    virtual void GetCurrencySymbolPosition(wxCurrencySymbolPosition& position, bool& hasSeparator) const = 0;
```
Sorry, I hadn't realized that splitting the enum in 2 would lead to using output parameters. If the choice is between using an enum with 4 values and the version with the output parameters, I think the enum with 4 values is better.

If you decide to add a struct containing both fields, I think calling it "position info" would be wrong as separator is not part of the position. Maybe just `wxCurrencySymbolPlacement`?

utelle committed:
Revert to ctors, and to 4-value symbol position

utelle commented:
@vadz 
Obviously,  using just `wxCFRef<NSNumberFormatter*> formatter = [[NSNumberFormatter alloc] init];` doesn't work. All mac builds fail with compile time errors. The compiler complains about unaccessible members. So, would something like the following work?

```C++
wxCFRef<NSNumberFormatter*> cfRefFormatter = [[NSNumberFormatter alloc] init];
NSNumberFormatter* formatter = cfRefFormatter;
```

utelle committed:
Attempt to fix macOS code

utelle committed:
Update documentation

vadz commented:
Sorry, just to check: are there any more changes planned here?

utelle commented:
> Sorry, just to check: are there any more changes planned here?

Not really. IMHO the added wxUILocale methods allow to retrieve the locale-dependent information relevant for formatting ordinary numeric values and currency values. Additionally, wxNumberFormatter was enhanced to apply locale-dependent grouping rules and to format currency values.

Regarding wxNumberFormatter there are of course conceivable extensions, but from my point of view they would make sense only, if wxNumberFormatter is a class and not just a bunch of static methods. Certainly not anything to be added to this PR.

The only point to clarify is whether the documentation I added is enough.

vadz reviewed (commented):
Thanks again for all this work, I'm pretty happy with the API now and most of the remaining remarks are trivial, but I'm worried by `RemoveCurrency()` which I had somehow overlooked before. Could you please explain if I'm missing something or stop using/calling it if I don't?

utelle committed:
Apply changes requested in review

utelle committed:
Make RemoveCurrencySymbolOrCode public

The function removes the currency symbol or code, and the grouping separators, according to the specified style flags.

utelle committed:
Add missing closing parenthesis
